src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -1549,7 +1549,7 @@ private static BoundBlock BindMethodBody(MethodSymbol method, TypeCompilationSta
                     binder = new ExecutableCodeBinder(arrowExpression, sourceMethod, binder);
                     importChain = binder.ImportChain;
                     // Add locals
-                    return binder.BindExpressionBodyAsBlock(arrowExpression, diagnostics);
+                    return binder.WithPatternVariablesIfAny(arrowExpression.Expression).BindExpressionBodyAsBlock(arrowExpression, diagnostics);
```

Comment:
If you look at the change made in src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs, you'll notice that I explicitly removed the behavior that you are suggesting. The reason is that the binder must be available to the SemanticModel in order for it (the model) to understand the scope tracked by the binder. Creating and dropping the binder on the fly doesn't work well in this case.